src/gui/guiFormSpecMenu.cpp

```diff
@@ -4228,153 +4286,384 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
-				if (s_count && button != BET_WHEEL_UP) {
-					// Non-empty stack has been clicked: select or shift-move it
+				if (button == BET_MIDDLE)
+					craft_amount = 10;
+				else if (event.MouseInput.Shift && button == BET_LEFT)
+					// TODO: We should craft everything with shift-left-click,
+					// but the slow crafting code limits us, so we only craft one
+					craft_amount = 1;
+					//craft_amount = list_s->getItem(s.i).getStackMax(m_client->idef());
+				else
+					craft_amount = 1;
+
+				// Holding shift moves the crafted item to the inventory
+				m_shift_craft = event.MouseInput.Shift;
+
+			} else if (!m_selected_item && button != BET_WHEEL_UP && !empty) {
+				// Non-empty stack has been clicked: select or shift-move it
+				u32 count;
+				if (button == BET_RIGHT)
+					count = (s_count + 1) / 2;
+				else if (button == BET_MIDDLE)
+					count = MYMIN(s_count, 10);
+				else if (button == BET_WHEEL_DOWN)
+					count = 1;
+				else if (button == BET_LEFT)
+					count = s_count;
+
+				if (event.MouseInput.Shift) {
+					// Shift pressed: move item, right click moves 1
+					shift_move_amount = button == BET_RIGHT ? 1 : count;
+				} else {
+					// No shift: select item
 					m_selected_item = new GUIInventoryList::ItemSpec(s);
+					m_selected_amount = count;
+					m_selected_dragging = button != BET_WHEEL_DOWN;
+				}
 
-					u32 count;
-					if (button == BET_RIGHT)
-						count = (s_count + 1) / 2;
-					else if (button == BET_MIDDLE)
-						count = MYMIN(s_count, 10);
-					else if (button == BET_WHEEL_DOWN)
-						count = 1;
-					else  // left
-						count = s_count;
-
-					if (!event.MouseInput.Shift) {
-						// no shift: select item
-						m_selected_amount = count;
-						m_selected_dragging = button != BET_WHEEL_DOWN;
-						m_auto_place = false;
-					} else {
-						// shift pressed: move item, right click moves 1
-						shift_move_amount = button == BET_RIGHT ? 1 : count;
+			} else if (m_selected_item) {
+				// Clicked a slot: move
+				if (button == BET_RIGHT || button == BET_WHEEL_UP)
+					move_amount = 1;
+				else if (button == BET_WHEEL_DOWN)
+					pickup_amount = MYMIN(s_count, 1);
+				else if (button == BET_MIDDLE)
+					move_amount = MYMIN(m_selected_amount, 10);
+				else if (button == BET_WHEEL_UP)
+					move_amount = 1;
+				else if (button == BET_LEFT)
+					move_amount = m_selected_amount;
+
+				if (event.MouseInput.Shift && !identical && matching) {
+					// Shift-move all of the selected item
+					move_amount = 0;
+
+					// Try to find somewhere to move the items to
+					u32 rings = m_inventory_rings.size();
+					u32 r = 0;
+					for (; r < rings; r++) {
+						const ListRingSpec &lr = m_inventory_rings[r];
+						if (lr.inventoryloc == s.inventoryloc &&
+								lr.listname == s.listname)
+							break;
 					}
-				}
-			} else { // m_selected_item != NULL
-				assert(m_selected_amount >= 1);
+					if (r >= rings) // Not found
+						break;
+					r = (r + 1) % rings;
```

Comment:
Would you please be so nice to move this list ring code to a separate function? That would help to keep the code a bit better structured.